File: src/test/java/pl/training/module8/incremental/MigratingPricingEngineTest.java
package pl.training.module8.incremental;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertInstanceOf;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.concurrent.atomic.AtomicInteger;

import org.junit.jupiter.api.Test;

final class MigratingPricingEngineTest {
    private static final PriceRequest REQUEST = new PriceRequest(
            new BigDecimal("100.00"), 1, BigDecimal.ZERO);
    private static final PriceQuote LEGACY_QUOTE = new PriceQuote(
            new BigDecimal("100.00"));
    private static final PriceQuote CANDIDATE_QUOTE = new PriceQuote(
            new BigDecimal("99.00"));

    @Test
    void legacyModeCallsOnlyTheLegacyImplementation() {
        var legacyCalls = new AtomicInteger();
        var candidateCalls = new AtomicInteger();
        var events = new ArrayList<VerificationEvent>();
        PricingEngine engine = engine(
                returning(legacyCalls, LEGACY_QUOTE),
                returning(candidateCalls, CANDIDATE_QUOTE),
                events::add,
                MigrationMode.LEGACY);

        PriceQuote result = engine.quote(REQUEST);

        assertEquals(LEGACY_QUOTE, result);
        assertEquals(1, legacyCalls.get());
        assertEquals(0, candidateCalls.get());
        assertEquals(0, events.size());
    }

    @Test
    void candidateModeCallsOnlyTheCandidateImplementation() {
        var legacyCalls = new AtomicInteger();
        var candidateCalls = new AtomicInteger();
        var events = new ArrayList<VerificationEvent>();
        PricingEngine engine = engine(
                returning(legacyCalls, LEGACY_QUOTE),
                returning(candidateCalls, CANDIDATE_QUOTE),
                events::add,
                MigrationMode.CANDIDATE);

        PriceQuote result = engine.quote(REQUEST);

        assertEquals(CANDIDATE_QUOTE, result);
        assertEquals(0, legacyCalls.get());
        assertEquals(1, candidateCalls.get());
        assertEquals(0, events.size());
    }

    @Test
    void verifyModeReportsAgreementAndReturnsTheLegacyResult() {
        var legacyCalls = new AtomicInteger();
        var candidateCalls = new AtomicInteger();
        var events = new ArrayList<VerificationEvent>();
        PricingEngine engine = engine(
                returning(legacyCalls, LEGACY_QUOTE),
                returning(candidateCalls, LEGACY_QUOTE),
                events::add,
                MigrationMode.VERIFY);

        PriceQuote result = engine.quote(REQUEST);

        assertSame(LEGACY_QUOTE, result);
        assertEquals(1, legacyCalls.get());
        assertEquals(1, candidateCalls.get());
        var agreement = assertInstanceOf(
                VerificationEvent.Agreement.class, events.get(0));
        assertEquals(REQUEST, agreement.request());
        assertEquals(LEGACY_QUOTE, agreement.quote());
    }

    @Test
    void verifyModeReportsDivergenceButStillReturnsTheLegacyResult() {
        var events = new ArrayList<VerificationEvent>();
        PricingEngine engine = engine(
                request -> LEGACY_QUOTE,
                request -> CANDIDATE_QUOTE,
                events::add,
                MigrationMode.VERIFY);

        PriceQuote result = engine.quote(REQUEST);

        assertSame(LEGACY_QUOTE, result);
        var divergence = assertInstanceOf(
                VerificationEvent.Divergence.class, events.get(0));
        assertEquals(REQUEST, divergence.request());
        assertEquals(LEGACY_QUOTE, divergence.legacyQuote());
        assertEquals(CANDIDATE_QUOTE, divergence.candidateQuote());
    }

    @Test
    void candidateFailureCannotChangeTheVerifyResponse() {
        var events = new ArrayList<VerificationEvent>();
        PricingEngine engine = engine(
                request -> LEGACY_QUOTE,
                request -> {
                    throw new IllegalStateException("candidate unavailable");
                },
                events::add,
                MigrationMode.VERIFY);

        PriceQuote result = engine.quote(REQUEST);

        assertSame(LEGACY_QUOTE, result);
        var failure = assertInstanceOf(
                VerificationEvent.CandidateFailure.class, events.get(0));
        assertEquals(REQUEST, failure.request());
        assertEquals(LEGACY_QUOTE, failure.legacyQuote());
        assertEquals("java.lang.IllegalStateException",
                failure.exceptionType());
        assertEquals("candidate unavailable", failure.message());
    }

    @Test
    void jvmErrorsAreNotMaskedByVerification() {
        var candidateFailure = new AssertionError("candidate corrupted");
        PricingEngine engine = engine(
                request -> LEGACY_QUOTE,
                request -> {
                    throw candidateFailure;
                },
                VerificationReporter.ignoring(),
                MigrationMode.VERIFY);

        AssertionError result = assertThrows(
                AssertionError.class,
                () -> engine.quote(REQUEST));

        assertSame(candidateFailure, result);
    }

    @Test
    void invalidCandidateResultIsReportedAsFailure() {
        var events = new ArrayList<VerificationEvent>();
        PricingEngine engine = engine(
                request -> LEGACY_QUOTE,
                request -> null,
                events::add,
                MigrationMode.VERIFY);

        PriceQuote result = engine.quote(REQUEST);

        assertSame(LEGACY_QUOTE, result);
        var failure = assertInstanceOf(
                VerificationEvent.CandidateFailure.class, events.get(0));
        assertEquals("java.lang.NullPointerException",
                failure.exceptionType());
        assertEquals("candidate quote", failure.message());
    }

    @Test
    void reporterFailureCannotChangeTheVerifyResponse() {
        PricingEngine engine = engine(
                request -> LEGACY_QUOTE,
                request -> LEGACY_QUOTE,
                event -> {
                    throw new IllegalStateException("reporter unavailable");
                },
                MigrationMode.VERIFY);

        assertSame(LEGACY_QUOTE, engine.quote(REQUEST));
    }

    @Test
    void legacyFailureRemainsAuthoritativeAndSkipsTheCandidate() {
        var candidateCalls = new AtomicInteger();
        var events = new ArrayList<VerificationEvent>();
        var legacyFailure = new IllegalStateException("legacy unavailable");
        PricingEngine engine = engine(
                request -> {
                    throw legacyFailure;
                },
                returning(candidateCalls, CANDIDATE_QUOTE),
                events::add,
                MigrationMode.VERIFY);

        IllegalStateException result = assertThrows(
                IllegalStateException.class,
                () -> engine.quote(REQUEST));

        assertSame(legacyFailure, result);
        assertEquals(0, candidateCalls.get());
        assertEquals(0, events.size());
    }

    @Test
    void missingRequestNeverReachesEitherImplementation() {
        var legacyCalls = new AtomicInteger();
        var candidateCalls = new AtomicInteger();
        PricingEngine engine = engine(
                returning(legacyCalls, LEGACY_QUOTE),
                returning(candidateCalls, CANDIDATE_QUOTE),
                VerificationReporter.ignoring(),
                MigrationMode.VERIFY);

        NullPointerException failure = assertThrows(
                NullPointerException.class,
                () -> engine.quote(null));

        assertEquals("request", failure.getMessage());
        assertEquals(0, legacyCalls.get());
        assertEquals(0, candidateCalls.get());
    }

    @Test
    void constructorRejectsEveryMissingDependency() {
        PricingEngine validEngine = request -> LEGACY_QUOTE;
        VerificationReporter validReporter = VerificationReporter.ignoring();

        assertMissing("legacy", () -> engine(
                null, validEngine, validReporter, MigrationMode.LEGACY));
        assertMissing("candidate", () -> engine(
                validEngine, null, validReporter, MigrationMode.LEGACY));
        assertMissing("reporter", () -> engine(
                validEngine, validEngine, null, MigrationMode.LEGACY));
        assertMissing("mode", () -> engine(
                validEngine, validEngine, validReporter, null));
    }

    private static MigratingPricingEngine engine(
            PricingEngine legacy,
            PricingEngine candidate,
            VerificationReporter reporter,
            MigrationMode mode) {
        return new MigratingPricingEngine(
                legacy, candidate, reporter, mode);
    }

    private static PricingEngine returning(
            AtomicInteger calls,
            PriceQuote result) {
        return request -> {
            calls.incrementAndGet();
            return result;
        };
    }

    private static void assertMissing(String message, Runnable action) {
        NullPointerException failure = assertThrows(
                NullPointerException.class, action::run);
        assertEquals(message, failure.getMessage());
    }
}

File: typescript/test/module8/incremental/MigratingPricingEngine.test.ts
import { AssertionError } from 'node:assert';
import { Decimal } from 'decimal.js';
import { describe, expect, it } from 'vitest';
import { IllegalStateError, NullPointerError } from '../../../src/shared/errors.js';
import { MigratingPricingEngine } from '../../../src/module8/incremental/MigratingPricingEngine.js';
import { MigrationMode } from '../../../src/module8/incremental/MigrationMode.js';
import { PriceQuote } from '../../../src/module8/incremental/PriceQuote.js';
import { PriceRequest } from '../../../src/module8/incremental/PriceRequest.js';
import type { PricingEngine } from '../../../src/module8/incremental/PricingEngine.js';
import {
  Agreement,
  CandidateFailure,
  Divergence,
  type VerificationEvent,
} from '../../../src/module8/incremental/VerificationEvent.js';
import { VerificationReporter } from '../../../src/module8/incremental/VerificationReporter.js';

// Odpowiednik AtomicInteger z testu Javy.
interface Counter {
  value: number;
}

const REQUEST = new PriceRequest(new Decimal('100.00'), 1, new Decimal(0));
const LEGACY_QUOTE = new PriceQuote(new Decimal('100.00'));
const CANDIDATE_QUOTE = new PriceQuote(new Decimal('99.00'));

function engine(
  legacy: PricingEngine,
  candidate: PricingEngine,
  reporter: VerificationReporter,
  mode: MigrationMode,
): MigratingPricingEngine {
  return new MigratingPricingEngine(legacy, candidate, reporter, mode);
}

function returning(calls: Counter, result: PriceQuote): PricingEngine {
  return {
    quote: () => {
      calls.value++;
      return result;
    },
  };
}

function constant(result: PriceQuote): PricingEngine {
  return { quote: () => result };
}

function collecting(events: VerificationEvent[]): VerificationReporter {
  return { report: (event) => events.push(event) };
}

function thrownBy(action: () => unknown): unknown {
  try {
    action();
  } catch (error) {
    return error;
  }
  throw new Error('expected an exception');
}

function assertMissing(message: string, action: () => unknown): void {
  const failure = thrownBy(action);
  expect(failure).toBeInstanceOf(NullPointerError);
  expect((failure as Error).message).toBe(message);
}

describe('MigratingPricingEngineTest', () => {
  it('legacyModeCallsOnlyTheLegacyImplementation', () => {
    const legacyCalls = { value: 0 };
    const candidateCalls = { value: 0 };
    const events: VerificationEvent[] = [];
    const pricingEngine: PricingEngine = engine(
      returning(legacyCalls, LEGACY_QUOTE),
      returning(candidateCalls, CANDIDATE_QUOTE),
      collecting(events),
      MigrationMode.LEGACY,
    );

    const result = pricingEngine.quote(REQUEST);

    expect(result.equals(LEGACY_QUOTE)).toBe(true);
    expect(legacyCalls.value).toBe(1);
    expect(candidateCalls.value).toBe(0);
    expect(events.length).toBe(0);
  });

  it('candidateModeCallsOnlyTheCandidateImplementation', () => {
    const legacyCalls = { value: 0 };
    const candidateCalls = { value: 0 };
    const events: VerificationEvent[] = [];
    const pricingEngine: PricingEngine = engine(
      returning(legacyCalls, LEGACY_QUOTE),
      returning(candidateCalls, CANDIDATE_QUOTE),
      collecting(events),
      MigrationMode.CANDIDATE,
    );

    const result = pricingEngine.quote(REQUEST);

    expect(result.equals(CANDIDATE_QUOTE)).toBe(true);
    expect(legacyCalls.value).toBe(0);
    expect(candidateCalls.value).toBe(1);
    expect(events.length).toBe(0);
  });

  it('verifyModeReportsAgreementAndReturnsTheLegacyResult', () => {
    const legacyCalls = { value: 0 };
    const candidateCalls = { value: 0 };
    const events: VerificationEvent[] = [];
    const pricingEngine: PricingEngine = engine(
      returning(legacyCalls, LEGACY_QUOTE),
      returning(candidateCalls, LEGACY_QUOTE),
      collecting(events),
      MigrationMode.VERIFY,
    );

    const result = pricingEngine.quote(REQUEST);

    expect(result).toBe(LEGACY_QUOTE);
    expect(legacyCalls.value).toBe(1);
    expect(candidateCalls.value).toBe(1);
    const agreement = events[0];
    expect(agreement).toBeInstanceOf(Agreement);
    if (!(agreement instanceof Agreement)) return;
    expect(agreement.request.equals(REQUEST)).toBe(true);
    expect(agreement.quote.equals(LEGACY_QUOTE)).toBe(true);
  });

  it('verifyModeReportsDivergenceButStillReturnsTheLegacyResult', () => {
    const events: VerificationEvent[] = [];
    const pricingEngine: PricingEngine = engine(
      constant(LEGACY_QUOTE),
      constant(CANDIDATE_QUOTE),
      collecting(events),
      MigrationMode.VERIFY,
    );

    const result = pricingEngine.quote(REQUEST);

    expect(result).toBe(LEGACY_QUOTE);
    const divergence = events[0];
    expect(divergence).toBeInstanceOf(Divergence);
    if (!(divergence instanceof Divergence)) return;
    expect(divergence.request.equals(REQUEST)).toBe(true);
    expect(divergence.legacyQuote.equals(LEGACY_QUOTE)).toBe(true);
    expect(divergence.candidateQuote.equals(CANDIDATE_QUOTE)).toBe(true);
  });

  it('candidateFailureCannotChangeTheVerifyResponse', () => {
    const events: VerificationEvent[] = [];
    const pricingEngine: PricingEngine = engine(
      constant(LEGACY_QUOTE),
      {
        quote: () => {
          throw new IllegalStateError('candidate unavailable');
        },
      },
      collecting(events),
      MigrationMode.VERIFY,
    );

    const result = pricingEngine.quote(REQUEST);

    expect(result).toBe(LEGACY_QUOTE);
    const failure = events[0];
    expect(failure).toBeInstanceOf(CandidateFailure);
    if (!(failure instanceof CandidateFailure)) return;
    expect(failure.request.equals(REQUEST)).toBe(true);
    expect(failure.legacyQuote.equals(LEGACY_QUOTE)).toBe(true);
    expect(failure.exceptionType).toBe('IllegalStateError');
    expect(failure.message).toBe('candidate unavailable');
  });

  it('jvmErrorsAreNotMaskedByVerification', () => {
    const candidateFailure = new AssertionError({ message: 'candidate corrupted' });
    const pricingEngine: PricingEngine = engine(
      constant(LEGACY_QUOTE),
      {
        quote: () => {
          throw candidateFailure;
        },
      },
      VerificationReporter.ignoring(),
      MigrationMode.VERIFY,
    );

    const result = thrownBy(() => pricingEngine.quote(REQUEST));

    expect(result).toBeInstanceOf(AssertionError);
    expect(result).toBe(candidateFailure);
  });

  it('invalidCandidateResultIsReportedAsFailure', () => {
    const events: VerificationEvent[] = [];
    const pricingEngine: PricingEngine = engine(
      constant(LEGACY_QUOTE),
      { quote: () => null as unknown as PriceQuote },
      collecting(events),
      MigrationMode.VERIFY,
    );

    const result = pricingEngine.quote(REQUEST);

    expect(result).toBe(LEGACY_QUOTE);
    const failure = events[0];
    expect(failure).toBeInstanceOf(CandidateFailure);
    if (!(failure instanceof CandidateFailure)) return;
    expect(failure.exceptionType).toBe('NullPointerError');
    expect(failure.message).toBe('candidate quote');
  });

  it('reporterFailureCannotChangeTheVerifyResponse', () => {
    const pricingEngine: PricingEngine = engine(
      constant(LEGACY_QUOTE),
      constant(LEGACY_QUOTE),
      {
        report: () => {
          throw new IllegalStateError('reporter unavailable');
        },
      },
      MigrationMode.VERIFY,
    );

    expect(pricingEngine.quote(REQUEST)).toBe(LEGACY_QUOTE);
  });

  it('legacyFailureRemainsAuthoritativeAndSkipsTheCandidate', () => {
    const candidateCalls = { value: 0 };
    const events: VerificationEvent[] = [];
    const legacyFailure = new IllegalStateError('legacy unavailable');
    const pricingEngine: PricingEngine = engine(
      {
        quote: () => {
          throw legacyFailure;
        },
      },
      returning(candidateCalls, CANDIDATE_QUOTE),
      collecting(events),
      MigrationMode.VERIFY,
    );

    const result = thrownBy(() => pricingEngine.quote(REQUEST));

    expect(result).toBeInstanceOf(IllegalStateError);
    expect(result).toBe(legacyFailure);
    expect(candidateCalls.value).toBe(0);
    expect(events.length).toBe(0);
  });

  it('missingRequestNeverReachesEitherImplementation', () => {
    const legacyCalls = { value: 0 };
    const candidateCalls = { value: 0 };
    const pricingEngine: PricingEngine = engine(
      returning(legacyCalls, LEGACY_QUOTE),
      returning(candidateCalls, CANDIDATE_QUOTE),
      VerificationReporter.ignoring(),
      MigrationMode.VERIFY,
    );

    assertMissing('request', () => pricingEngine.quote(null as unknown as PriceRequest));
    expect(legacyCalls.value).toBe(0);
    expect(candidateCalls.value).toBe(0);
  });

  it('constructorRejectsEveryMissingDependency', () => {
    const validEngine = constant(LEGACY_QUOTE);
    const validReporter = VerificationReporter.ignoring();

    assertMissing('legacy', () =>
      engine(null as unknown as PricingEngine, validEngine, validReporter, MigrationMode.LEGACY),
    );
    assertMissing('candidate', () =>
      engine(validEngine, null as unknown as PricingEngine, validReporter, MigrationMode.LEGACY),
    );
    assertMissing('reporter', () =>
      engine(validEngine, validEngine, null as unknown as VerificationReporter, MigrationMode.LEGACY),
    );
    assertMissing('mode', () =>
      engine(validEngine, validEngine, validReporter, null as unknown as MigrationMode),
    );
  });
});
